xaitk_saliency/impls/saliency/fsal.py

```diff
@@ -0,0 +1,65 @@
+from xaitk_saliency import ImageClassifierSaliencyMapGenerator
+from xaitk_saliency.utils.masking import weight_regions_by_scalar
+
+import numpy as np
+
+
+class Fsal (ImageClassifierSaliencyMapGenerator):
+    """
+    This saliency implementation transforms black-box
+    image classification scores into saliency heatmaps. This should
+    require a sequence of per-class confidences predicted on the
+    reference image, a number of per-class confidences as predicted
+    on perturbed images, as well as the masks of the reference image
+    perturbations (as would be output from a
+    `PerturbImage` implementation.
+    """
+
+    def generate(
+            self,
+            image_conf: np.ndarray,
+            perturbed_conf: np.ndarray,
+            perturbed_masks: np.ndarray
+    ) -> np.ndarray:
+        """
+        Generate an image saliency heat-map matrix given a the original classifier
+        confidence on image, confidences on pertrubed images and finally the masks
+        used to pertrube the image.
+
+        :param image_conf:
+            Numpy Original classifier confidence on Image for all classes
+            that require saliency map generation: [nClasses], float, [0, 1] range
+        :param perturbed_conf:
+            Numpy Classifier confidences on pertrubed version of Image
+            for nClasses: [nMasks x nClasses], float, [0, 1] range
+        :param perturbed_masks:
+            Numpy Array of masks used to pertrube Image in the same order
+            as their their respective 'pertutbed_conf's: [nMasks x H x W],
+            float, [0, 1] range.
+
+        :return: Numpy array of heat-map matrices for each input class in nClasses.
+
+            :py:class:`numpy.ndarray` matrix of the shape [nClasses, height, width]
+            shape as the input image matrix but of floating-point type within
+            the range of [0,1], where areas of higher value represent more
+            salient regions according to the classifier that generated confidence
+            values.
+        """
+
+        # Creating an emtpy array of length nClasses and shape H X W
+        sal = np.empty((len(image_conf), *perturbed_masks.shape[-2:]))
+
+        # Iterating through each class confidence and compare it with
+        # its perturbed twin
+        for i, base_class_conf in enumerate(image_conf):
+            diff = np.empty(len(perturbed_conf[:, i]))
+            for j in range(len(perturbed_conf[:, i])):
+                diff[j] = base_class_conf - perturbed_conf[j][i]
```

Comment:
I feel like this can be vectorized (maybe with some broadcasting)? Looping over each class does not seem to be the most efficient way to do this.

Reply:
Good point, I vectorized this to perform saliency computation across all classes and regions of the image.